File: tasty-mima/src/main/scala/tastymima/Problem.scala
package tastymima

import tastyquery.Modifiers.*
import tastyquery.Names.*
import tastyquery.Signatures.*
import tastyquery.Types.*

import tastymima.intf.{ProblemKind, Problem as IProblem}

final class Problem(val kind: ProblemKind, val path: List[Name], val details: Matchable) extends IProblem:
  import Problem.*

  def this(kind: ProblemKind, path: List[Name]) = this(kind, path, ())

  val pathString: String =
    val s1 = path.mkString(".")
    if path.nonEmpty && path.last.isTypeName && path.last.toTypeName.wrapsObjectName then s1 + "$"
    else s1

  def getKind(): ProblemKind = kind

  def getPathString(): String = pathString

  override def getDescription(): String | Null =
    (kind, details) match
      // Better sentence structure than putting `details` at the end for some combinations

      case (ProblemKind.MissingTermMember, details: Signature) =>
        s"The member ${getPathString()} with signature $details does not have a correspondant in current version"
      case (ProblemKind.NewAbstractMember, details: Signature) =>
        s"The member ${getPathString()} with signature $details "
          + "was concrete or did not exist but is abstract in current version"

      case _ =>
        val superDesc = super.getDescription()

        if details == () then superDesc
        else s"$superDesc: ${detailsString(details)}"
  end getDescription

  private def detailsString(details: Matchable): String = details match
    case details: TypeMappable => details.showBasic
    case details: OpenLevel    => openLevelToString(details)
    case details: BeforeAfter  => s"before: ${detailsString(details.before)}; after: ${detailsString(details.after)}"
    case _                     => details.toString()
  end detailsString

  override def toString(): String =
    s"Problem($kind, $pathString)"
end Problem

object Problem:
  /** Used as `details` for a `Problem` when there is something to show "before" and "after". */
  final case class BeforeAfter(before: Matchable, after: Matchable)

  private def openLevelToString(level: OpenLevel): String = level match
    case OpenLevel.Final  => "final"
    case OpenLevel.Sealed => "sealed"
    case OpenLevel.Closed => "(default)"
    case OpenLevel.Open   => "open"
  end openLevelToString
end Problem

File: tasty-mima-interface/src/main/java/tastymima/intf/Problem.java
package tastymima.intf;

public interface Problem {
  public ProblemKind getKind();
  public String getPathString();

  public default String getDescription() {
    String path = getPathString();
    switch (getKind()) {
      case MissingClass:
        return "The class " + path + " does not have a correspondant in current version";
      case MissingTypeMember:
        return "The type member " + path + " does not have a correspondant in current version";
      case MissingTermMember:
        return "The member " + path + " does not have a correspondant in current version";
      case RestrictedVisibilityChange:
        return "The symbol " + path + " has a more restrictive visibility qualifier in current version";
      case IncompatibleKindChange:
        return "The symbol " + path + " has an incompatible kind in current version";
      case MissingParent:
        return "The class " + path + " is missing a parent in current version";
      case IncompatibleSelfTypeChange:
        return "The class " + path + " has an incompatible self type in current version";
      case RestrictedOpenLevelChange:
        return "The class " + path + " has a more restrictive open level (open, sealed, final) in current version";
      case AbstractClass:
        return "The class " + path + " was concrete but is abstract in current version";
      case FinalMember:
        return "The member " + path + " was open but is final in current version";
      case TypeArgumentCountMismatch:
        return "The class " + path + " does not have the same number of type arguments in current version";
      case IncompatibleTypeChange:
        return "The symbol " + path + " has an incompatible type in current version";
      case NewAbstractMember:
        return "The member " + path + " was concrete or did not exist but is abstract in current version";
      case InternalError:
        return "Internal error while checking symbol " + path;
      default:
        return "Unknown problem " + getKind() + " about symbol " + path;
    }
  }

  public default String getFilterIncantation() {
    return "ProblemMatcher.make(ProblemKind." + getKind() + ", \"" + getPathString() + "\")";
  }
}
